### ISSUE TITLE
DeprecationWarnings with Python 3.8

### ISSUE THREAD

**Issue by newAM** [ronf/asyncssh#234]
Using python 3.8 with asyncssh results in significant numbers of deprecation warnings.

> "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead

**Comment by ronf** [comment]
Thanks for the report. See the comments in #196 - the issue is that this new syntax is fundamentally incompatible with Python 3.4, which AsyncSSH still supports. While I fully support code using AsyncSSH using the new async syntax, and even have conditionals which add things like async iterators in the later Python versions which can handle them, the implementation of AsyncSSH is still built using only "yield from" and @asyncio.coroutine, so that it works on Python 3.4.

I'll probably drop the older syntax in favor of async/await and make the minimum supported Python version either 3.5 or 3.6 in AsyncSSH 2.0, but so far I haven't had the time to do that update, and unfortunately I don't know of an easy way to globally silence the deprecation warnings in the meantime.

**Comment by newAM** [comment]
The syntax changes do make things difficult!

I took a peek with grep and counted:

- 2,520 `yield from`
- 603 `@asyncio.coroutine`
- 2 `if python344` conditions
- 6 `if python35` conditions
- 1 `if python352` condition

I got introduced to asyncio in 3.6 so I am still unfamiliar with the older syntax, but I'm willing to take a naive attempt at updating!  Do you think there will be any other (obvious) problems with updating?

**Comment by ronf** [comment]
I appreciate the offer, but the syntax changes aren't the main issue. This can largely be done in an automated fashion. I'll probably want to use this as an opportunity to make some other changes at the same time, though, removing some deprecated functionality and perhaps taking advantage of some Python features that only appeared in post-3.4 Python. I just need to spend some time figuring out what those should be.

Now that 3.8 is officially out and 3.4 is officially end-of-life, it's probably about time I get started on this. I'll see if I can make a first cut (at least as deciding what changes I want to make) by this coming weekend.

**Comment by newAM** [comment]
Let me know if I can help at all!  I have gotten a lot of use out of asyncssh and I appreciate all the effort that has gone into it!

**Comment by ronf** [comment]
Thanks very much! Once I have something working, it'd be great if you could help with testing it.

A quick update: I have all the syntax converted over and I've pulled out the conditionals related to Python versions that will no longer be needed. My plan right now is to require a minimum of Python 3.6, to avoid some incompatible changes in context managers between Python 3.5 and 3.5.2 that I previously had to work around.

I've also made a bunch of changes to take better advantage of async context managers ("async with") in the unit tests, eliminating the need for explicit calls to methods like wait_closed().

I've got a clean pylint run on the updated code, but it's not quite working yet. I'm currently running through the unit tests to figure out what's wrong.

I'm testing with Python 3.8 for now, and that's also showing some other deprecations that I will need to address.

**Comment by newAM** [comment]
Wow that was quick!  Testing I can definitely do, let me know when there's a changeset ready to consume and I'll drop it into an existing project and give it a go.

**Comment by Ch3LL** [comment]
I'm more than willing to help test when these changes are pushed as well :) I'm excited for the move to the new syntax. thanks!

**Comment by ronf** [comment]
Thanks, Megan - it's getting there, and I'll reply back here when I have something ready to test.

Just to let you know, you can already use the new async/await syntax in any code which calls into AsyncSSH. That has been supported for quite some time (since back in the 3.5 days when that syntax was first added). I even added support for things like async iterators and async context managers within AsyncSSH, despite the fact that the implementation of those features was actually still written using the old "yield from" syntax to allow a single code base to work for pre & post Python 3.5 (with a small number of conditional version checks).

The only thing changing here is that AsyncSSH itself will be using the async/await syntax, which should make the code a bit nicer to read, and avoid a bunch of deprecation warnings.

**Comment by ronf** [comment]
Ok - I have a first cut at a version of AsyncSSH which uses the new async/await syntax. Since some of the async API was still evolving in Python 3.5, I decided to make this release require a minimum of Python 3.6. You can find a source tarball for this initial version at:

    https://www.timeheart.net/asyncssh-2.0.0a1.tar.gz

This version also removes a few things that were previously marked as deprecated in earlier AsyncSSH releases. Full documentation will be available about this when I put together the official release.

This version also drops support for non-async context manager being used with await ("with await ..."), allowing only "async with" now. Using "async with" is much better, as it can actually wait asynchronously for things to properly close when the context manager exits, which the non-async version couldn't do. That could only start the close process, but you had to call "wait_closed" yourself to actually wait for the close to finish.

Some other syntax deprecated by Python has also been cleaned up, such as the use of "async with" on lock objects and removing the "loop" parameter from some calls.

Feedback is extremely welcome!

**Comment by newAM** [comment]
I did some surface level testing and found an interesting print in python3.8 using the example from the docs:
```python
import asyncio, asyncssh, sys

async def run_client():
    async with asyncssh.connect('localhost') as conn:
        result = await conn.run('echo "Hello!"', check=True)
        print(result.stdout, end='')

try:
    asyncio.get_event_loop().run_until_complete(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit('SSH connection failed: ' + str(exc))
```

3.7
```bash
Hello!
```

3.8
```
An open stream object is being garbage collected; call "stream.close()" explicitly.
Hello!
```

...I did some research and only found the issue that you reported: https://bugs.python.org/issue38529

Apart from those warnings everything seems to be working well!

**Comment by ronf** [comment]
Thanks for helping to test this!

As you saw, I ran into the same issue here in my testing, and was able to confirm that this is a new check in Python 3.8 that seems like it doesn't properly handle the garbage collector running after you call close() but before connection_lost() is triggered by the close being acknowledged by the peer.

I'm hoping this can be addressed in the asyncio code, but I may experiment a bit to see if I can suppress this error in AsyncSSH in the meantime. For now, it seems harmless other than the noise it causes in the output, but I'd like to avoid people thinking there's something wrong with their code, as I did when I first saw it.

**Comment by ronf** [comment]
Unfortunately, it looks like suppressing this error would require setting the event loop's global exception handler, which I'm not sure a library like AsyncSSH should be doing. That configuration really belongs to the application code which sets up the event loop.

For now, I think I'll check this code into the "develop" branch but hold off a bit on doing the 2.0 release while I wait to hear back on the bug I filed.

I have some other stuff I want to get in before I do a release anyway, such as moving to cryptography 2.8 so I can pick up the implementation of Chacha20-Poly1305 it now has and remove my one remaining dependency on libnacl.

**Comment by ronf** [comment]
After a bit of thought, I came up with a way to address the Python 3.8 warnings about not closing streams. It's a bit ugly, but it seems to work. Basically, if you await on wait_closed() on a stream before you allow the reader/writer to be garbage collected, the Python 3.8 error won't appear. However, wait_closed() was only added in 3.7, and I want AsyncSSH to work in 3.6. 3.6 doesn't output this warning, though, so wait_closed() isn't needed there. So, I created a maybe_wait_closed() function that will attempt to call wait_closed() on a stream, but ignore the error if that method doesn't exist, and I call that in the places where just calling close() without waiting was triggering the false error.

With this change, I'm now getting a clean unittest run again. In fact, I cleaned up some other warnings that were being output as well (reducing their log level from warning to debug1), so that AsyncSSH doesn't output any logging to the terminal by default.

I also added the ability to use the return value from connect_agent() as an async context manager, and I added wait_closed() as a method on SSHWriter, mirroring what was done in asyncio.StreamWriter in Python 3.7 (though the SSHWriter method works even in 3.6).

All of these changes are now in the "develop" branch.

**Comment by newAM** [comment]
That does seem like the best solution, makes it cleaner for the implementation side which is always nice.  Thank you for all the effort you put into this!

**Comment by ronf** [comment]
Thanks! Application code which calls close() without calling wait_closed() might still run into this error, but wait_closed() is now available to call on SSHAgentClient even on Python 3.6 (using my "maybe" wrapper), and using "async with" is an even nicer way to write such code. For instance:

```python
    async with asyncssh.connect_agent() as agent:
        await agent.add_keys([key1, key2])
```

**Comment by ronf** [comment]
These changes are now released as part of AsyncSSH 2.0.0. Complete details of the changes can be found at https://asyncssh.readthedocs.io/en/latest/changes.html. Thanks for the help testing it!
